Sprint 20. Homework

DoNatPanic opened the pull request:
Привет!
В файле RetrofitClient я написала две функции doRequest и doRequestById. Сделала по аналогии с примером проекта с фильмами (о котором идет речь в теории на курсе). 
Не уверена, что это хорошее решение, так как данные функции отличаются только одной строчкой и не универсальны. А как сделать лучше - не придумала. Буду признательна за подсказку

DoNatPanic committed:
Sprint 20. Homework. Part 1: update debounce with coroutines

Выполнен первый пункт задания:

1. Переделайте реализацию debounce для отправки поискового запроса и нажатия на элементы списка. Вместо прежнего подхода с использованием Handler используйте корутины. Это позволит сделать процесс более эффективным и управляемым.

DoNatPanic committed:
Sprint 20. Homework. Part 2: update the implementation of the playback progress display

Выполнен второй пункт задания:

2. Обновите реализацию отображения прогресса воспроизведения на экране «Аудиоплеер». Теперь в этой реализации нужно использовать корутины. Это обеспечит более простое и надёжное отображение прогресса.

DoNatPanic committed:
Sprint 20. Homework. Part 3: coroutines in working with the network

Выполнен второй пункт задания:

3. выполнение самого сетевого запроса для поиска треков теперь должно быть реализовано с использованием корутин и Flow на всех уровнях архитектуры исходного кода

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/data/network/RetrofitApi.kt:
```diff
@@ -1,14 +1,13 @@
 
     @GET("/lookup?entity=song")
-    fun searchById(@Query("id") trackId: Long): Call<TracksResponse>
+    suspend fun searchById(@Query("id") trackId: Long): TracksResponse
```
А зачем вообще этот метод нужен? Информацию о треке на экран плеера можно передавать через аргументы при навигации, данных, которые приходят в поиске, для этого достаточно. Рекомендую убрать лишний метод и воспользоваться `Parcelable` для передачи информации активити при навигации

DoNatPanic replied:
Применила Parcelable, не знала о таком инструменте.
Надеюсь все правильно написала, прошу проверить и поделиться, если можно сделать лучше или если что-то сделано неверно.

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/data/network/TracksRepositoryImpl.kt:
```diff
@@ -3,14 +3,35 @@ package com.example.playlistmaker.data.network
+            }
+
+            else -> {
+                emit(ApiResponse.Success(response))
```
Для сетевого взаимодействия следует делать отдельную модель, а не пользоваться domain моделью. Лучше создать класс TrackDto и добавить маппинг сюда в репозиторий

DoNatPanic replied:
Сделала, надеюсь что все верно расфасовала по папкам

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/ui/audioplayer/view_model/PlayerViewModel.kt:
```diff
@@ -20,103 +23,104 @@ class PlayerViewModel(
     fun elapsedTimeLiveData(): LiveData<Long> = elapsedTimeState
 
     init {
         val getTrackDetailsUseCase: GetTrackDetailsUseCase = getKoin().get()
```
Все зависимости вьюмодель должна получать в конструкторе, на это есть требование. Это позволяет вьюмодели оставаться тестируемой и не зависеть от коина

DoNatPanic replied:
Сделано

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/ui/audioplayer/view_model/PlayerViewModel.kt:
```diff
@@ -20,103 +23,104 @@ class PlayerViewModel(
-        if (playerState.value == PlayerState.PLAYING) {
-            tracksInteractor.pause()
+    private fun startTimer() {
+        timerJob = viewModelScope.launch {
```
При запуске новой джобы на всякий случай лучше отменять предыдущую:
```
timerJob?.cancel()
timerJob = ...
```

DoNatPanic replied:
Добавила, спасибо за замечание

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/ui/audioplayer/view_model/PlayerViewModel.kt:
```diff
@@ -20,103 +23,104 @@ class PlayerViewModel(
     }
 
+    private fun timerFlow(elTime: Long): Flow<Long> = flow {
+        (elTime until PREVIEW_TIME step DELAY).forEach {
+            delay(DELAY)
+            emit(it)
+        }
+    }.flowOn(Dispatchers.IO)
```
Не стоит такой таймер делать, аппаратные счетчики не являются идеальными и `delay(1000)` не обязательно задержит выполнение именно на 1000мс, может быть погрешность. Для отображения времени воспроизведения трека необходимо использовать `MediaPlayer::getCurrentPosition`

Плюс IO здесь сильно лишний, т.к. он нужен для долгих блокирующих операций, а тут их нет. Можно использовать Default вместо IO

DoNatPanic replied:
Если честно, не совсем поняла, как это провернуть. Мне не понятно как вызывать getCurrentPosition, ведь это надо делать у экземпляра MediaPlayer. А что означает запись `MediaPlayer::getCurrentPosition` не совсем понятно..

ArturNurtdinov replied:
1. Это действительно нужно делать у экземпляра `MediaPlayer`, тебе нужно для этого сделать функцию в твоём интеракторе, которая будет вызывать аналогичную функцию репозитория, который в свою очередь вызовет `getCurrentPosition`, возвращаемым результатом будет Long
2. Запись `MediaPlayer::getCurrentPosition` является ссылкой на функцию - член класса `MediaPlayer`

DoNatPanic replied:
Спасибо, исправила

DoNatPanic committed:
Fix critical problems

1. Все данные, передаваемые между слоями Data и Domain, должны проходить через DTO
2. SearchViewModel и PlayerViewModel используют функции getKoin().get() для создания зависимостей, а не получают их в конструкторе
3. Прогресс воспроизведения обновляется раз в 100мс

DoNatPanic committed:
Apply reviewer's recomendations

DoNatPanic committed:
Apply reviewer's recomendations 2

1. Убрали зависимость вью модели от коина
2. Добавлен timerJob?.cancel() перед вызовом джобы

DoNatPanic commented:
В комментарии коммита `Fix critical problems` сказано что устранена зависимость вьюмодели от коина, фактически это произошло в коммите `Apply reviewer's recomendations 2`

DoNatPanic committed:
Apply reviewer's recomendations 3

Вместо аппаратного счетчика используется MediaPlayer::getCurrentPosition
